### ISSUE TITLE
Lower priority (or change order) of include directory?

### ISSUE THREAD

**Issue by dethtoll1** [Microsoft/vcpkg#4122]
Hello,

tldr version: 
I'm looking for a reliable way to re-order include directories so that the vcpkg include directory appears *after* other directories (using CMake).

Longer version:
I am integrating vcpkg into a sizeable CMake-based codebase.  I'm finally at the point where I have everything working, but am having trouble ensuring that our codebase will build on all machines going forward.  

This codebase has a separate repository of external dependencies - those that are not yet in vcpkg, or need a specific version/customization (let's call these "customized dependencies")  We would like the include folders for those dependencies to be processed *before* the vcpkg include folder.  

Why?  Well, if we have a customized dependency that also exists in vcpkg (let's use boost as an example), then "#include <boost/...>" would have different results depending on whether the vcpkg boost is installed or not (and depending on the order of inherited dependencies in CMake).  

It seems that the order of include directories inherited from interface targets is difficult to control in CMake.  Carefully re-ordering our target dependencies to make the vcpkg-based targets appear last would be difficult and error-prone.  I'm looking for a more robust solution that's transparent to most of our developers.

Thanks!

**Comment by ras0219-msft** [comment]
Thanks for the question!

In this case, I would strongly recommend using a specific vcpkg instance for this project that _only_ contains the dependencies you want from vcpkg and does not contain _any_ of the customized dependencies.

If you need to customize a dependency that is used by something else -- for example, you want to use cpprestsdk from vcpkg but want to customize boost -- then you should push your customizations into vcpkg. Otherwise, you're potentially violating ODR and could get really nasty crashes at runtime!

To push customizations into vcpkg is quite easy, just create a patch file and add it to the `ports/xyz/portfile.cmake` either via a call to `vcpkg_apply_patches()` or the `PATCHES` parameter to `vcpkg_from_github()`. I'd be happy to go into more detail here if you encounter any issues!

**Comment by dethtoll1** [comment]
Thanks for the quick reply! 

I think that pushes me in the right direction.  Is there some mechanism to ensure that a vcpkg only contains a specified list of dependencies (and their transitive dependencies)?  

Our CMake process automatically clones vcpkg (if not present) at a specific changeset and installs a list of N packages (if not installed).  But there's no guard against extra packages.  I can obviously check for the N packages, but would ideally not need to maintain an 'approved' list of all transitive dependencies.

Maybe there's a vcpkg command that lists all dependencies for a given package?

**Comment by degski** [comment]
@ras0219-msft 

I see this as a problem as well, you're explanation, though, of how this is all quite easy is beyond me. F..e. "just create a patch", Hmmm . . . What I mean is that going into detail for the uninitiated is probably very useful, but I would not do it in this issue as it will be closed and will disappear of the radar, the wiki is probably a better place. And... examples always do miracles ...

**Comment by ras0219-msft** [comment]
@degski Here's a full example about how to produce patches for a library in vcpkg: https://github.com/Microsoft/vcpkg/blob/master/docs/examples/patching-libpng.md. A "patch" is a standard format for representing changes to source files[1].

@dethtoll1 I'm not sure exactly what you'd want here, however `vcpkg list` will print out a list of every currently installed package. You could additional run `vcpkg remove <x> <y> <z>` to ensure that x y and z are _not_ installed.

To get the dependencies for a given package, you can look in `ports\<p>\CONTROL` which has a field `Build-Depends:` listing all immediate dependencies. However, I think you would care more about the transitive dependency list and we don't currently have a command to print that out (other than `vcpkg install --dry-run <x>` in a clean enlistment).

[1] https://en.wikipedia.org/wiki/Patch_(Unix)

**Comment by degski** [comment]
@ras0219-msft Thanks for the pointer.

**Comment by meitiever** [comment]
Do not patch the vcpkg, because you will forget how and why to patch it when you deliver the software to the customer.
the way to reference the customized headers before vcpkg in visual studio is to add your headers to "Additional Include Directories", not "include directories". 
the path defined in the first variable has a higher priority than the vcpkg. and solve my problem perfectly.
this may help.
